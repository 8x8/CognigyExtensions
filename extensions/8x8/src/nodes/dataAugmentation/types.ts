import type { INodeFunctionBaseParams } from '@cognigy/extension-tools';
import type StoreLocationName from '../../constants/StoreLocationName';
import type { I8x8SimpleConnection } from '../../connections/8x8SimpleConnection';
import type { SearchGenericParams } from '../../types';
import type { DataAugmentationRequestBody } from './api/postDataAugmentation';

export interface DataAugmentationParams extends SearchGenericParams {
  displayName1: string
  value1: string
  displayName2: string
  value2: string
  displayName3: string
  value3: string
  displayName4: string
  value4: string
  displayName5: string
  value5: string
}

export interface IDataAugmentationParams extends INodeFunctionBaseParams {
  config: {
    connection: I8x8SimpleConnection
    sipCallId: string
    useCustomFields: boolean
    customFields: DataAugmentationRequestBody
    storeLocation: StoreLocationName
    contextKey?: string
    inputKey?: string
  } & DataAugmentationParams
}
