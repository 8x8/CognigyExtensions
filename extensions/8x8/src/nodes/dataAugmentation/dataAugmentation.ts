import { createNodeDescriptor } from '@cognigy/extension-tools';
import type { INodeField } from '@cognigy/extension-tools/build/interfaces/descriptor';
import { defaultJSONProperties } from './constants/defaultJSONProperties';
import dataAugmentationResolver from './resolvers/dataAugmentationResolver';
import {
  DISPLAY_NAME_DESCRIPTION,
  DISPLAY_NAME_TITLE,
  VALUE_FIELD_TITLE,
  VALUE_FIELD_DESCRIPTION
} from './constants/labels';

const augmentationFieldsSection1: INodeField[] = [
  {
    key: 'displayName1',
    label: DISPLAY_NAME_TITLE,
    description: DISPLAY_NAME_DESCRIPTION,
    type: 'cognigyText'
  },
  {
    key: 'value1',
    label: VALUE_FIELD_TITLE,
    description: VALUE_FIELD_DESCRIPTION,
    type: 'cognigyText'
  }

];

const augmentationFieldsSection2: INodeField[] = [
  {
    key: 'displayName2',
    label: DISPLAY_NAME_TITLE,
    description: DISPLAY_NAME_DESCRIPTION,
    type: 'cognigyText'
  },
  {
    key: 'value2',
    label: VALUE_FIELD_TITLE,
    description: VALUE_FIELD_DESCRIPTION,
    type: 'cognigyText'
  }

];

const augmentationFieldsSection3: INodeField[] = [
  {
    key: 'displayName3',
    label: DISPLAY_NAME_TITLE,
    description: DISPLAY_NAME_DESCRIPTION,
    type: 'cognigyText'
  },
  {
    key: 'value3',
    label: VALUE_FIELD_TITLE,
    description: VALUE_FIELD_DESCRIPTION,
    type: 'cognigyText'
  }

];

const augmentationFieldsSection4: INodeField[] = [
  {
    key: 'displayName4',
    label: DISPLAY_NAME_TITLE,
    description: DISPLAY_NAME_DESCRIPTION,
    type: 'cognigyText'
  },
  {
    key: 'value4',
    label: VALUE_FIELD_TITLE,
    description: VALUE_FIELD_DESCRIPTION,
    type: 'cognigyText'
  }

];

const augmentationFieldsSection5: INodeField[] = [
  {
    key: 'displayName5',
    label: DISPLAY_NAME_TITLE,
    description: DISPLAY_NAME_DESCRIPTION,
    type: 'cognigyText'
  },
  {
    key: 'value5',
    label: VALUE_FIELD_TITLE,
    description: VALUE_FIELD_DESCRIPTION,
    type: 'cognigyText'
  }
];

export const dataAugmentationNode = createNodeDescriptor({
  type: 'dataAugmentation',
  defaultLabel: 'Data Augmentation',
  summary: 'Sends a payload with variables to enrich the properties served to the agent',
  fields: [
    {
      key: 'connection',
      label: '8x8 Connection',
      type: 'connection',
      params: {
        connectionType: 'eightbyeightsimple',
        required: true
      }
    },
    {
      key: 'sipCallId',
      label: '8x8 sip Call ID',
      description: 'The ID of the call',
      type: 'cognigyText',
      params: {
        required: true
      }
    },
    ...augmentationFieldsSection1,
    ...augmentationFieldsSection2,
    ...augmentationFieldsSection3,
    ...augmentationFieldsSection4,
    ...augmentationFieldsSection5,
    {
      key: 'useCustomFields',
      type: 'toggle',
      label: {
        default: 'Use JSON Fields to configure payload'
      },
      description: {
        default: 'Use JSON for defining payload for Data Augmentation.'
      },
      defaultValue: false
    },
    {
      key: 'customFields',
      label: '8x8 Data Augmentation JSON properties',
      type: 'json',
      description: 'Send 8x8 properties with custom JSON for Data Augmentation payload, default value provided for payload structure model.',
      defaultValue: defaultJSONProperties,
      params: {
        required: true
      },
      condition: {
        key: 'useCustomFields',
        value: true
      }
    },
    {
      key: 'storeLocation',
      type: 'select',
      label: 'Where to store the result',
      defaultValue: 'input',
      params: {
        options: [
          {
            label: 'Input',
            value: 'input'
          },
          {
            label: 'Context',
            value: 'context'
          }
        ],
        required: true
      }
    },
    {
      key: 'inputKey',
      type: 'cognigyText',
      label: 'Input Key to store Result',
      defaultValue: 'scheduleStatus',
      condition: {
        key: 'storeLocation',
        value: 'input'
      }
    },
    {
      key: 'contextKey',
      type: 'cognigyText',
      label: {
        default: 'Context Key to store Result'
      },
      defaultValue: 'scheduleStatus',
      condition: {
        key: 'storeLocation',
        value: 'context'
      }
    }
  ],
  sections: [
    {
      key: 'augmentationSection1',
      label: 'Data Augmentation payload value section 1',
      defaultCollapsed: false,
      fields: [
        ...(augmentationFieldsSection1.map(field => field.key))
      ],
      condition: {
        key: 'useCustomFields',
        value: false
      }
    },
    {
      key: 'augmentationSection2',
      label: 'Data Augmentation payload value section 2',
      defaultCollapsed: false,
      fields: [
        ...(augmentationFieldsSection2.map(field => field.key))
      ],
      condition: {
        key: 'useCustomFields',
        value: false
      }
    },
    {
      key: 'augmentationSection3',
      label: 'Data Augmentation payload value section 3',
      defaultCollapsed: false,
      fields: [
        ...(augmentationFieldsSection3.map(field => field.key))
      ],
      condition: {
        key: 'useCustomFields',
        value: false
      }
    },
    {
      key: 'augmentationSection4',
      label: 'Data Augmentation payload value section 4',
      defaultCollapsed: false,
      fields: [
        ...(augmentationFieldsSection4.map(field => field.key))
      ],
      condition: {
        key: 'useCustomFields',
        value: false
      }
    },
    {
      key: 'augmentationSection5',
      label: 'Data Augmentation payload value section 5',
      defaultCollapsed: false,
      fields: [
        ...(augmentationFieldsSection5.map(field => field.key))
      ],
      condition: {
        key: 'useCustomFields',
        value: false
      }
    },
    {
      key: 'jsonPayload',
      label: 'Data Augmentation JSON payload',
      defaultCollapsed: false,
      fields: [
        'customFields'
      ]
    },
    {
      key: 'storage',
      label: 'Storage Option',
      defaultCollapsed: true,
      fields: [
        'storeLocation',
        'inputKey',
        'contextKey'
      ]
    }
  ],
  form: [
    { type: 'field', key: 'connection' },
    { type: 'field', key: 'sipCallId' },
    { type: 'field', key: 'useCustomFields' },
    { type: 'section', key: 'augmentationSection1' },
    { type: 'section', key: 'augmentationSection2' },
    { type: 'section', key: 'augmentationSection3' },
    { type: 'section', key: 'augmentationSection4' },
    { type: 'section', key: 'augmentationSection5' },
    { type: 'section', key: 'jsonPayload' },
    { type: 'section', key: 'storage' }
  ],
  appearance: {
    color: '#ff0050'
  },
  function: dataAugmentationResolver

});
