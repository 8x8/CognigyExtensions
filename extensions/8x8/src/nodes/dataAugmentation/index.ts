import type { INodeDescriptor } from '@cognigy/extension-tools';
import { dataAugmentationNode } from './dataAugmentation';

export const getDataAugmentationNode = (): INodeDescriptor[] => ([
  dataAugmentationNode
]);
