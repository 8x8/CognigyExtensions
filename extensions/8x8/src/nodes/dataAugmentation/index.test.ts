import { getDataAugmentationNode } from '../dataAugmentation/index';

describe('voice handover node > getDataAugmentationNode', () => {
  it('should create the node for data augmentation', () => {
    const nodes = getDataAugmentationNode();

    expect(nodes).toMatchSnapshot();
  });
});
