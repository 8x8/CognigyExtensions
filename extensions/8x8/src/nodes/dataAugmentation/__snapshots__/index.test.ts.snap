// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`voice handover node > getDataAugmentationNode should create the node for data augmentation 1`] = `
[
  {
    "_id": undefined,
    "appearance": {
      "color": "#ff0050",
      "showIcon": true,
      "textColor": "black",
      "variant": "regular",
    },
    "behavior": {
      "entrypoint": false,
      "stopping": false,
    },
    "constraints": {
      "collapsable": false,
      "creatable": true,
      "deletable": true,
      "editable": true,
      "movable": true,
      "placement": {
        "children": {
          "whitelist": [],
        },
        "predecessor": {},
        "successor": {},
      },
    },
    "defaultLabel": "Data Augmentation",
    "dependencies": {
      "children": [],
    },
    "fields": [
      {
        "key": "connection",
        "label": "8x8 Connection",
        "params": {
          "connectionType": "eightbyeightsimple",
          "required": true,
        },
        "type": "connection",
      },
      {
        "description": "The ID of the call",
        "key": "sipCallId",
        "label": "8x8 sip Call ID",
        "params": {
          "required": true,
        },
        "type": "cognigyText",
      },
      {
        "description": "The name of the property to be sent via payload and displayed for the Agent",
        "key": "displayName1",
        "label": "Title of the property",
        "type": "cognigyText",
      },
      {
        "description": "The value of the property to be sent via payload and displayed for the Agent",
        "key": "value1",
        "label": "The value of the property",
        "type": "cognigyText",
      },
      {
        "description": "The name of the property to be sent via payload and displayed for the Agent",
        "key": "displayName2",
        "label": "Title of the property",
        "type": "cognigyText",
      },
      {
        "description": "The value of the property to be sent via payload and displayed for the Agent",
        "key": "value2",
        "label": "The value of the property",
        "type": "cognigyText",
      },
      {
        "description": "The name of the property to be sent via payload and displayed for the Agent",
        "key": "displayName3",
        "label": "Title of the property",
        "type": "cognigyText",
      },
      {
        "description": "The value of the property to be sent via payload and displayed for the Agent",
        "key": "value3",
        "label": "The value of the property",
        "type": "cognigyText",
      },
      {
        "description": "The name of the property to be sent via payload and displayed for the Agent",
        "key": "displayName4",
        "label": "Title of the property",
        "type": "cognigyText",
      },
      {
        "description": "The value of the property to be sent via payload and displayed for the Agent",
        "key": "value4",
        "label": "The value of the property",
        "type": "cognigyText",
      },
      {
        "description": "The name of the property to be sent via payload and displayed for the Agent",
        "key": "displayName5",
        "label": "Title of the property",
        "type": "cognigyText",
      },
      {
        "description": "The value of the property to be sent via payload and displayed for the Agent",
        "key": "value5",
        "label": "The value of the property",
        "type": "cognigyText",
      },
      {
        "defaultValue": false,
        "description": {
          "default": "Use JSON for defining payload for Data Augmentation.",
        },
        "key": "useCustomFields",
        "label": {
          "default": "Use JSON Fields to configure payload",
        },
        "type": "toggle",
      },
      {
        "condition": {
          "key": "useCustomFields",
          "value": true,
        },
        "defaultValue": {
          "data": {
            "variables": [
              {
                "display": "true",
                "displayName": "mock name to be displayed 1",
                "ivr": "true",
                "name": "mockName1",
                "value": "mock value 1",
              },
              {
                "display": "true",
                "displayName": "mock name to be displayed 2",
                "ivr": "true",
                "name": "mockName2",
                "value": "mock value 2",
              },
            ],
          },
        },
        "description": "Send 8x8 properties with custom JSON for Data Augmentation payload, default value provided for payload structure model.",
        "key": "customFields",
        "label": "8x8 Data Augmentation JSON properties",
        "params": {
          "required": true,
        },
        "type": "json",
      },
      {
        "defaultValue": "input",
        "key": "storeLocation",
        "label": "Where to store the result",
        "params": {
          "options": [
            {
              "label": "Input",
              "value": "input",
            },
            {
              "label": "Context",
              "value": "context",
            },
          ],
          "required": true,
        },
        "type": "select",
      },
      {
        "condition": {
          "key": "storeLocation",
          "value": "input",
        },
        "defaultValue": "scheduleStatus",
        "key": "inputKey",
        "label": "Input Key to store Result",
        "type": "cognigyText",
      },
      {
        "condition": {
          "key": "storeLocation",
          "value": "context",
        },
        "defaultValue": "scheduleStatus",
        "key": "contextKey",
        "label": {
          "default": "Context Key to store Result",
        },
        "type": "cognigyText",
      },
    ],
    "form": [
      {
        "key": "connection",
        "type": "field",
      },
      {
        "key": "sipCallId",
        "type": "field",
      },
      {
        "key": "useCustomFields",
        "type": "field",
      },
      {
        "key": "augmentationSection1",
        "type": "section",
      },
      {
        "key": "augmentationSection2",
        "type": "section",
      },
      {
        "key": "augmentationSection3",
        "type": "section",
      },
      {
        "key": "augmentationSection4",
        "type": "section",
      },
      {
        "key": "augmentationSection5",
        "type": "section",
      },
      {
        "key": "jsonPayload",
        "type": "section",
      },
      {
        "key": "storage",
        "type": "section",
      },
    ],
    "function": [Function],
    "parentType": null,
    "preview": undefined,
    "sections": [
      {
        "condition": {
          "key": "useCustomFields",
          "value": false,
        },
        "defaultCollapsed": false,
        "fields": [
          "displayName1",
          "value1",
        ],
        "key": "augmentationSection1",
        "label": "Data Augmentation payload value section 1",
      },
      {
        "condition": {
          "key": "useCustomFields",
          "value": false,
        },
        "defaultCollapsed": false,
        "fields": [
          "displayName2",
          "value2",
        ],
        "key": "augmentationSection2",
        "label": "Data Augmentation payload value section 2",
      },
      {
        "condition": {
          "key": "useCustomFields",
          "value": false,
        },
        "defaultCollapsed": false,
        "fields": [
          "displayName3",
          "value3",
        ],
        "key": "augmentationSection3",
        "label": "Data Augmentation payload value section 3",
      },
      {
        "condition": {
          "key": "useCustomFields",
          "value": false,
        },
        "defaultCollapsed": false,
        "fields": [
          "displayName4",
          "value4",
        ],
        "key": "augmentationSection4",
        "label": "Data Augmentation payload value section 4",
      },
      {
        "condition": {
          "key": "useCustomFields",
          "value": false,
        },
        "defaultCollapsed": false,
        "fields": [
          "displayName5",
          "value5",
        ],
        "key": "augmentationSection5",
        "label": "Data Augmentation payload value section 5",
      },
      {
        "defaultCollapsed": false,
        "fields": [
          "customFields",
        ],
        "key": "jsonPayload",
        "label": "Data Augmentation JSON payload",
      },
      {
        "defaultCollapsed": true,
        "fields": [
          "storeLocation",
          "inputKey",
          "contextKey",
        ],
        "key": "storage",
        "label": "Storage Option",
      },
    ],
    "summary": "Sends a payload with variables to enrich the properties served to the agent",
    "tags": [],
    "tokens": [],
    "type": "dataAugmentation",
  },
]
`;
